api/oneclick.py

```diff
@@ -155,16 +170,157 @@ def circulate_item(self, patron_id, item_id, return_item=False):
+        """Place a book on hold.
+
+        :param patron: a Patron object for the patron who wants to check out the book.
+        :param pin: The patron's password (not used).
+        :param licensepool: The Identifier of the book to be checked out is 
+        attached to this licensepool.
+        :param internal_format: Represents the patron's desired book format.  Ignored for now.
+
+        :return: A HoldInfo object on success, None on failure
+        """
+        patron_oneclick_id = self.validate_patron(patron)
+        (item_oneclick_id, item_media) = self.validate_item(licensepool)
+
+        resp_dict = self.circulate_item(patron_id=patron_oneclick_id, item_id=item_oneclick_id, hold=True, return_item=False)
+
+        if not resp_dict or ('error_code' in resp_dict):
+            return None
+
+        self.log.debug("Patron %s/%s checked out item %s with transaction id %s.", patron.authorization_identifier, 
+            patron_oneclick_id, item_oneclick_id, resp_dict['transactionId'])
+
+        today = datetime.datetime.now()
+
+        hold = HoldInfo(
+            identifier_type=licensepool.identifier.type,
+            identifier=item_oneclick_id,
+            start_date=today,
+            # OneClick sets hold expirations to 2050-12-31, as a "forever"
+            end_date=None,
+            hold_position=0,
+        )
+
+        return hold
+
+
+    def release_hold(self, patron, pin, licensepool):
+        """Release a patron's hold on a book.
+
+        :param patron: a Patron object for the patron who wants to return the book.
+        :param pin: The patron's password (not used).
+        :param licensepool: The Identifier of the book to be checked out is 
+        attached to this licensepool.
+
+        :return True on success, raises circulation exceptions on failure.
+        """
+        patron_oneclick_id = self.validate_patron(patron)
+        (item_oneclick_id, item_media) = self.validate_item(licensepool)
+
+        resp_dict = self.circulate_item(patron_id=patron_oneclick_id, item_id=item_oneclick_id, hold=True, return_item=True)
+
+        if resp_dict == {}:
+            self.log.debug("Patron %s/%s released hold %s.", patron.authorization_identifier, 
+                patron_oneclick_id, item_oneclick_id)
+            return True
+
+        # should never happen
+        raise CirculationException("Unknown error %s/%s releasing %s.", patron.authorization_identifier, 
+            patron_oneclick_id, item_oneclick_id)
+
+
+
+    ''' -------------------------- Patron Account Handling -------------------------- '''
+    def create_patron(self, patron):
+        """ Ask OneClick to create a new patron record.
+
+        :param patron: a Patron object which contains the permanent id to give to 
+        OneClick as the patron's library card number.
+        :return OneClick's internal patron id.
+        """
+
+        patron_cardno = patron.authorization_identifier
```

Comment:
This is not going to work in all libraries, but I don't have a solution in place yet, so it's fine for now.